src/Servers.jl

```diff
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
-        sslconfig = SSLConfig(require_ssl_verification)
-    end
+    handler = f isa Handler ? f : stream ? StreamHandlerFunction(f) : RequestHandlerFunction(f)
 
-    @info "Listening on: $host"
-    if isassigned(tcpref)
+    inet = getinet(host, port)
+    if tcpref !== nothing
         tcpserver = tcpref[]
     elseif reuseaddr
         tcpserver = Sockets.TCPServer(; delay=false)
-        if Sys.islinux() || Sys.isapple()
+        if Sys.isunix()
             rc = ccall(:jl_tcp_reuseport, Int32, (Ptr{Cvoid},), tcpserver.handle)
-            Sockets.bind(tcpserver, host.host, host.port; reuseaddr=true)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         else
             @warn "reuseaddr=true may not be supported on this platform: $(Sys.KERNEL)"
-            Sockets.bind(tcpserver, host.host, host.port; reuseaddr=true)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         end
         Sockets.listen(tcpserver)
     else
-        tcpserver = Sockets.listen(host)
-        tcpref[] = tcpserver
+        tcpserver = Sockets.listen(inet)
     end
+    verbose && @info "Listening on: $host:$port"
 
-    if host isa Sockets.InetAddr # build debugging info
-        hostname = string(host.host)
-        hostport = string(host.port)
-    else
-        hostname = string(host)
-        hostport = ""
+    if tcpisvalid === nothing
+        tcpisvalid = ratelimit === nothing ? x->true : x->check_rate_limit(x, ratelimit)
     end
-    listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
+
+    return listenloop(handler, Server2(sslconfig, tcpserver, string(host), string(port)), tcpisvalid,
+        connectioncounter, reuse_limit, readtimeout, verbose)
 end
 
-function listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
-    try
-        id = 0
-        while isopen(tcpserver)
-            try
-                io = accept(tcpserver)
-                if !tcpisvalid(io; kw...)
-                    @info "Accept-Reject:  $io"
-                    close(io)
-                    continue
-                end
-                io = getsslcontext(io, sslconfig)
-                let i=id, conn = Connection(hostname, hostport, pipeline_limit, 0, require_ssl_verification, io)
-                    @async try
-                        @info "Accept ($i):  $conn"
-                        connectioncounter[] += 1
-                        handle_connection(f, conn; kw...)
+"main server loop that accepts new tcp connections and spawns async threads to handle them"
+function listenloop(h, server,
+    tcpisvalid=x->true, connectioncounter=Ref(0),
+    reuse_limit::Int=1, readtimeout::Int=0, verbose::Bool=false)
+    count = 1
+    while isopen(server)
+        try
+            io = accept(server)
+            io === nothing && continue
```

Comment:
Is this needed?
The`?Sockets.accept` spec says "returns a connection to the client.".
How can `accept` return `nothing`?

Reply:
I don't know how, but:
```julia
julia> t = Sockets.listen(parse(IPAddr, "127.0.0.1"), 8081)
Sockets.TCPServer(RawFD(0x00000014) active)

julia> @code_warntype acc(t)
Body::Union{Nothing, TCPSocket}
2 1 ─ %1 = Sockets.accept::Core.Compiler.Const(Sockets.accept, false)                                                                                             │
  │   %2 = invoke %1(_2::Sockets.TCPServer)::Union{Nothing, TCPSocket}                                                                                            │
  └──      return %2                                                                                                                                              │
```

Reply:
Doesn't that just mean "the compiler thinks there may be a case where nothing is returned" rather than "sometimes nothing is returned" ?

If the spec for `listen` says that it returns a connection, and we're worried that it does not always do that. Then we need to `@assert io isa TCPSocket` and report a bug upstream if it ever fails.